hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java

```diff
@@ -318,7 +320,7 @@ public boolean isMemberAllowedToJoin(Address address) {
         lock.lock();
         try {
             ClusterState clusterState = node.getClusterService().getClusterState();
-            if (clusterState == ClusterState.FROZEN || clusterState == ClusterState.PASSIVE) {
+            if (!clusterState.isMigrationAllowed()) {
```

Comment:
Still, I will add `clusterState != ClusterState.IN_TRANSITION`  check here too. To make it more visible.